python_modules/dagster/dagster/_daemon/backfill.py

```diff
@@ -1,19 +1,78 @@
 import logging
 import sys
-from typing import Iterable, Mapping, Optional, cast
+import threading
+from typing import TYPE_CHECKING, Iterable, Mapping, Optional, Set, cast
+
+import pendulum
 
 from dagster._core.execution.asset_backfill import execute_asset_backfill_iteration
 from dagster._core.execution.backfill import BulkActionStatus, PartitionBackfill
 from dagster._core.execution.job_backfill import execute_job_backfill_iteration
 from dagster._core.workspace.context import IWorkspaceProcessContext
-from dagster._utils.error import SerializableErrorInfo, serializable_error_info_from_exc_info
+from dagster._utils.error import serializable_error_info_from_exc_info
+
+if TYPE_CHECKING:
+    from .daemon import TDaemonGenerator
+
+
+ACTIVE_BACKFILLS_INTERVAL_SECONDS = 2
+
+
+def execute_backfill_iteration_loop(
+    workspace_process_context: IWorkspaceProcessContext,
+    logger: logging.Logger,
+    shutdown_event: threading.Event,
+    until=None,
+) -> "TDaemonGenerator":
+    """
+    Helper function that performs sensor evaluations on a tighter loop, while reusing grpc locations
+    within a given daemon interval.  Rather than relying on the daemon machinery to run the
+    iteration loop every 30 seconds, sensors are continuously evaluated, every 5 seconds. We rely on
+    each sensor definition's min_interval to check that sensor evaluations are spaced appropriately.
+    """
+    from .controller import DEFAULT_DAEMON_INTERVAL_SECONDS
+
+    active_backfills: Set[str] = set()
+    while True:
+        start_time = pendulum.now("UTC").timestamp()
+        if until and start_time >= until:
+            # provide a way of organically ending the loop to support test environment
+            break
+
+        for backfill in execute_backfill_iteration(workspace_process_context, logger):
+            if backfill:
+                if backfill.status.is_active:
+                    active_backfills.add(backfill.backfill_id)
```

Comment:
i am very down with the overall plan here - but I think given that the very first thing that execute_backfill_iteration does is fetch all requested backfills, we may be able to accomplish this goal without needing to pass anything back from the generator?

Could the logic be something like

- fetch all requested backfills in this function (instead of at the beginning of execute_backfill_iteration)
```
backfills = instance.get_backfills(status=BulkActionStatus.REQUESTED)
``` 
- pass those backfills into execute_backfill_iteration as an argument

- If there were any backfills in that list, wait ACTIVE_BACKFILLS_INTERVAL_SECONDS - if not wait DEFAULT_DAEMON_INTERVAL_SECONDS

The downside of that is you might do one last extra fast loop once there are no requested backfills, but that's not too expensive and it will short-circuit quickly

Reply:
you could also consider only doing the ACTIVE_BACKFILLS_INTERVAL_SECONDS if there are specifically active asset backfills? from the description sounds like its only useful for those

Reply:
I believe the majority of backfills (including asset backfills) will transition from REQUESTED to COMPLETED within the first `execute_backfill_iteration` that covers them. So we'd be doing roughly double the number of fast loops. But I agree that the loops will be very cheap when there are no requested backfills.

Is this suggestion about mainly code readability or performance?

Reply:
Response from Daniel that I received via email but somehow didn't make it on here: "I guess the former but not strongly held"

@prha - care to break the tie?

Reply:
I do think that the loops of loops is a bit tough to reason about, especially since we have branched logic between assets/jobs, encompassed in the same job pass.  May be out of the scope of this particular PR, but I'd much rather see a split daemon between job backfills / asset backfills, since we've now broken the constraint that backfills are completed after a successful iteration.

I'm not that worried about the rate of DB fetches, since we have tight fetches in other cases (e.g. a single backfill scheduling a large number of runs will also hit the DB in a tight loop to look for backfill cancels).